src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -1397,6 +1397,33 @@ namespace cryptonote
     return res;
   }
   //-----------------------------------------------------------------------------------------------
+
+  bool core::check_storage_server_ping() const
+  {
+    CRITICAL_REGION_LOCAL(m_ping_lock);
+
+    if (m_last_storage_server_ping.time_since_epoch().count() == 0) {
+      MWARNING("Never heard from the storage server!");
+      return false;
+    }
+
+    const auto elapsed = std::chrono::system_clock::now() - m_last_storage_server_ping;
+
+    if (elapsed > STORAGE_SERVER_PING_LIFETIME) {
+      MWARNING("Last heard from the storage server: "
+               << std::chrono::duration_cast<std::chrono::minutes>(elapsed).count() << " minutes ago!");
+      return false;
+    }
+
+    return true;
+  }
+  //-----------------------------------------------------------------------------------------------
+  void core::update_storage_server_last_ping()
+  {
+    CRITICAL_REGION_LOCAL(m_ping_lock);
```

Comment:
Not sure we care about races here. If there's a race here, the value derived from system_clock::now() is going to be within milliseconds.